{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 13 November 2019\n",
    "# Calculate recovery of ClinVar pathogenic/likely-pathogenic variants in RefSeq exonic regions (GRCh37) by slop\n",
    "### by Pavlos Bousounis\n",
    "***Last updated 11/14/2019***\n",
    "___"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Import modules"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 293,
   "metadata": {},
   "outputs": [],
   "source": [
    "from datetime import datetime\n",
    "import glob\n",
    "from natsort import natsorted\n",
    "import numpy as np\n",
    "import os\n",
    "import pandas as pd\n",
    "import pathlib\n",
    "import pybedtools\n",
    "from pybedtools import BedTool\n",
    "import re"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Set working directory"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 512,
   "metadata": {},
   "outputs": [],
   "source": [
    "basedir = '/Users/pbousounis/Experiments/2019-10-29_hg19mod/RefSeq-ClinVar_GRCh37_slop_region_recovery/data/'\n",
    "os.chdir(basedir)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Save working directory and date"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 296,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Currently in /Users/pbousounis/Experiments/2019-10-29_hg19mod/RefSeq-ClinVar_GRCh37_slop_region_recovery/data\n",
      "\n",
      "Today is 2019-11-14\n"
     ]
    }
   ],
   "source": [
    "# get current working directory\n",
    "basedir = os.getcwd()\n",
    "# get today's date as YYYY-MM-DD\n",
    "today = datetime.today().strftime('%Y-%m-%d')\n",
    "\n",
    "print('Currently in ' + basedir + '\\n')\n",
    "print('Today is ' + today)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Define function ***bed_overlap()***\n",
    "___"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 311,
   "metadata": {},
   "outputs": [],
   "source": [
    "def bed_overlap(ref_bed_filepath, test_bed_filepath, dir_out='.'):\n",
    "    \n",
    "    \"\"\" Given two bed files, ref_bed and test_bed, perform bedtools intersect -c to return the number of \n",
    "    test_bed regions that overlap any regions in the ref_bed file. Returns the ref_bed file with a column \n",
    "    of overlap counts \"\"\"\n",
    "    \n",
    "    cwd = os.getcwd()\n",
    "    \n",
    "    # specify the reference bed file\n",
    "    ref_bedtool = BedTool(ref_bed_filepath)\n",
    "    prfx_ref = ref_bed_filepath.split('/')[-1]\n",
    "    prfx_ref = prfx_ref.split('.')[0]\n",
    "    \n",
    "    # specify the new ClinVar bed file\n",
    "    test_bedtool = BedTool(test_bed_filepath)\n",
    "    prfx_test = test_bed_filepath.split('/')[-1]\n",
    "    prfx_test = prfx_test.split('.')[0]\n",
    "\n",
    "    # specify name/path of output bed file\n",
    "    bed_out = dir_out + '/{}_IN_{}.bed'.format(prfx_test, prfx_ref)\n",
    "    \n",
    "    # run bedtools intersect to get all test_bed regions NOT found in ref_bed (-v option)\n",
    "    ref_in_test = test_bedtool.intersect(b=ref_bedtool, c=True)\n",
    "    \n",
    "    return(ref_in_test)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Run bedtools slop on RefSeq bed \n",
    "\n",
    "#### For ***n*** in {5, 10, 15, .., 150}:\n",
    "    extend each input bed region by n bases on each end and save as bed to disk\n",
    "___"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 306,
   "metadata": {},
   "outputs": [],
   "source": [
    "# create/designate slop file output directory\n",
    "slop_dir_out = os.path.join((today + \"_RefSeqGRCh37_exon_slop_bed\"))\n",
    "pathlib.Path(slop_dir_out).mkdir(exist_ok=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 309,
   "metadata": {},
   "outputs": [],
   "source": [
    "# create the RefSeq bedtool\n",
    "refseq_bed_filepath = '2019-11-08_RefSeq-GRCh37_latest_genomicGFF3.bed'\n",
    "refseq_bed = pd.read_csv(refseq_bed_filepath, sep='\\t', header=None, low_memory=False)\n",
    "refseq_bed[0] = 'chr' + refseq_bed.loc[:, 0]    \n",
    "\n",
    "# create bedtool from dataframe\n",
    "refseq_bedtool = BedTool.from_dataframe(refseq_bed)\n",
    "\n",
    "# save headless refseq bed\n",
    "refseq_bed_out = '2019-11-08_RefSeq-GRCh37_latest_genomicGFF3_headless.bed'\n",
    "refseq_bed.to_csv(refseq_bed_out, sep='\\t', header=None, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 310,
   "metadata": {},
   "outputs": [],
   "source": [
    "# retrieve hg19 chromosome sizes for pybedtools slop\n",
    "hg19_genome = pybedtools.helpers.get_chromsizes_from_ucsc(genome='hg19', saveas='data/hg19.genome')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Success! File saved:\n",
      "/Users/pbousounis/Experiments/2019-10-29_hg19mod/RefSeq-ClinVar_GRCh37_slop_region_recovery/data/2019-11-13_RefSeqGRCh37_exon_slop_bed/2019-11-13_RefSeqGRCh37_slop5.bed\n",
      "\n",
      "\n",
      "Success! File saved:\n",
      "/Users/pbousounis/Experiments/2019-10-29_hg19mod/RefSeq-ClinVar_GRCh37_slop_region_recovery/data/2019-11-13_RefSeqGRCh37_exon_slop_bed/2019-11-13_RefSeqGRCh37_slop10.bed\n",
      "\n",
      "\n",
      "Success! File saved:\n",
      "/Users/pbousounis/Experiments/2019-10-29_hg19mod/RefSeq-ClinVar_GRCh37_slop_region_recovery/data/2019-11-13_RefSeqGRCh37_exon_slop_bed/2019-11-13_RefSeqGRCh37_slop15.bed\n",
      "\n",
      "\n",
      "Success! File saved:\n",
      "/Users/pbousounis/Experiments/2019-10-29_hg19mod/RefSeq-ClinVar_GRCh37_slop_region_recovery/data/2019-11-13_RefSeqGRCh37_exon_slop_bed/2019-11-13_RefSeqGRCh37_slop20.bed\n",
      "\n",
      "\n",
      "Success! File saved:\n",
      "/Users/pbousounis/Experiments/2019-10-29_hg19mod/RefSeq-ClinVar_GRCh37_slop_region_recovery/data/2019-11-13_RefSeqGRCh37_exon_slop_bed/2019-11-13_RefSeqGRCh37_slop25.bed\n",
      "\n",
      "\n",
      "Success! File saved:\n",
      "/Users/pbousounis/Experiments/2019-10-29_hg19mod/RefSeq-ClinVar_GRCh37_slop_region_recovery/data/2019-11-13_RefSeqGRCh37_exon_slop_bed/2019-11-13_RefSeqGRCh37_slop30.bed\n",
      "\n",
      "\n",
      "Success! File saved:\n",
      "/Users/pbousounis/Experiments/2019-10-29_hg19mod/RefSeq-ClinVar_GRCh37_slop_region_recovery/data/2019-11-13_RefSeqGRCh37_exon_slop_bed/2019-11-13_RefSeqGRCh37_slop35.bed\n",
      "\n",
      "\n",
      "Success! File saved:\n",
      "/Users/pbousounis/Experiments/2019-10-29_hg19mod/RefSeq-ClinVar_GRCh37_slop_region_recovery/data/2019-11-13_RefSeqGRCh37_exon_slop_bed/2019-11-13_RefSeqGRCh37_slop40.bed\n",
      "\n",
      "\n",
      "Success! File saved:\n",
      "/Users/pbousounis/Experiments/2019-10-29_hg19mod/RefSeq-ClinVar_GRCh37_slop_region_recovery/data/2019-11-13_RefSeqGRCh37_exon_slop_bed/2019-11-13_RefSeqGRCh37_slop45.bed\n",
      "\n",
      "\n",
      "Success! File saved:\n",
      "/Users/pbousounis/Experiments/2019-10-29_hg19mod/RefSeq-ClinVar_GRCh37_slop_region_recovery/data/2019-11-13_RefSeqGRCh37_exon_slop_bed/2019-11-13_RefSeqGRCh37_slop50.bed\n",
      "\n",
      "\n",
      "Success! File saved:\n",
      "/Users/pbousounis/Experiments/2019-10-29_hg19mod/RefSeq-ClinVar_GRCh37_slop_region_recovery/data/2019-11-13_RefSeqGRCh37_exon_slop_bed/2019-11-13_RefSeqGRCh37_slop55.bed\n",
      "\n",
      "\n",
      "Success! File saved:\n",
      "/Users/pbousounis/Experiments/2019-10-29_hg19mod/RefSeq-ClinVar_GRCh37_slop_region_recovery/data/2019-11-13_RefSeqGRCh37_exon_slop_bed/2019-11-13_RefSeqGRCh37_slop60.bed\n",
      "\n",
      "\n",
      "Success! File saved:\n",
      "/Users/pbousounis/Experiments/2019-10-29_hg19mod/RefSeq-ClinVar_GRCh37_slop_region_recovery/data/2019-11-13_RefSeqGRCh37_exon_slop_bed/2019-11-13_RefSeqGRCh37_slop65.bed\n",
      "\n",
      "\n",
      "Success! File saved:\n",
      "/Users/pbousounis/Experiments/2019-10-29_hg19mod/RefSeq-ClinVar_GRCh37_slop_region_recovery/data/2019-11-13_RefSeqGRCh37_exon_slop_bed/2019-11-13_RefSeqGRCh37_slop70.bed\n",
      "\n",
      "\n",
      "Success! File saved:\n",
      "/Users/pbousounis/Experiments/2019-10-29_hg19mod/RefSeq-ClinVar_GRCh37_slop_region_recovery/data/2019-11-13_RefSeqGRCh37_exon_slop_bed/2019-11-13_RefSeqGRCh37_slop75.bed\n",
      "\n",
      "\n",
      "Success! File saved:\n",
      "/Users/pbousounis/Experiments/2019-10-29_hg19mod/RefSeq-ClinVar_GRCh37_slop_region_recovery/data/2019-11-13_RefSeqGRCh37_exon_slop_bed/2019-11-13_RefSeqGRCh37_slop80.bed\n",
      "\n",
      "\n",
      "Success! File saved:\n",
      "/Users/pbousounis/Experiments/2019-10-29_hg19mod/RefSeq-ClinVar_GRCh37_slop_region_recovery/data/2019-11-13_RefSeqGRCh37_exon_slop_bed/2019-11-13_RefSeqGRCh37_slop85.bed\n",
      "\n",
      "\n",
      "Success! File saved:\n",
      "/Users/pbousounis/Experiments/2019-10-29_hg19mod/RefSeq-ClinVar_GRCh37_slop_region_recovery/data/2019-11-13_RefSeqGRCh37_exon_slop_bed/2019-11-13_RefSeqGRCh37_slop90.bed\n",
      "\n",
      "\n",
      "Success! File saved:\n",
      "/Users/pbousounis/Experiments/2019-10-29_hg19mod/RefSeq-ClinVar_GRCh37_slop_region_recovery/data/2019-11-13_RefSeqGRCh37_exon_slop_bed/2019-11-13_RefSeqGRCh37_slop95.bed\n",
      "\n",
      "\n",
      "Success! File saved:\n",
      "/Users/pbousounis/Experiments/2019-10-29_hg19mod/RefSeq-ClinVar_GRCh37_slop_region_recovery/data/2019-11-13_RefSeqGRCh37_exon_slop_bed/2019-11-13_RefSeqGRCh37_slop100.bed\n",
      "\n",
      "\n",
      "Success! File saved:\n",
      "/Users/pbousounis/Experiments/2019-10-29_hg19mod/RefSeq-ClinVar_GRCh37_slop_region_recovery/data/2019-11-13_RefSeqGRCh37_exon_slop_bed/2019-11-13_RefSeqGRCh37_slop105.bed\n",
      "\n",
      "\n",
      "Success! File saved:\n",
      "/Users/pbousounis/Experiments/2019-10-29_hg19mod/RefSeq-ClinVar_GRCh37_slop_region_recovery/data/2019-11-13_RefSeqGRCh37_exon_slop_bed/2019-11-13_RefSeqGRCh37_slop110.bed\n",
      "\n",
      "\n",
      "Success! File saved:\n",
      "/Users/pbousounis/Experiments/2019-10-29_hg19mod/RefSeq-ClinVar_GRCh37_slop_region_recovery/data/2019-11-13_RefSeqGRCh37_exon_slop_bed/2019-11-13_RefSeqGRCh37_slop115.bed\n",
      "\n",
      "\n",
      "Success! File saved:\n",
      "/Users/pbousounis/Experiments/2019-10-29_hg19mod/RefSeq-ClinVar_GRCh37_slop_region_recovery/data/2019-11-13_RefSeqGRCh37_exon_slop_bed/2019-11-13_RefSeqGRCh37_slop120.bed\n",
      "\n",
      "\n",
      "Success! File saved:\n",
      "/Users/pbousounis/Experiments/2019-10-29_hg19mod/RefSeq-ClinVar_GRCh37_slop_region_recovery/data/2019-11-13_RefSeqGRCh37_exon_slop_bed/2019-11-13_RefSeqGRCh37_slop125.bed\n",
      "\n",
      "\n",
      "Success! File saved:\n",
      "/Users/pbousounis/Experiments/2019-10-29_hg19mod/RefSeq-ClinVar_GRCh37_slop_region_recovery/data/2019-11-13_RefSeqGRCh37_exon_slop_bed/2019-11-13_RefSeqGRCh37_slop130.bed\n",
      "\n",
      "\n",
      "Success! File saved:\n",
      "/Users/pbousounis/Experiments/2019-10-29_hg19mod/RefSeq-ClinVar_GRCh37_slop_region_recovery/data/2019-11-13_RefSeqGRCh37_exon_slop_bed/2019-11-13_RefSeqGRCh37_slop135.bed\n",
      "\n",
      "\n",
      "Success! File saved:\n",
      "/Users/pbousounis/Experiments/2019-10-29_hg19mod/RefSeq-ClinVar_GRCh37_slop_region_recovery/data/2019-11-13_RefSeqGRCh37_exon_slop_bed/2019-11-13_RefSeqGRCh37_slop140.bed\n",
      "\n",
      "\n",
      "Success! File saved:\n",
      "/Users/pbousounis/Experiments/2019-10-29_hg19mod/RefSeq-ClinVar_GRCh37_slop_region_recovery/data/2019-11-13_RefSeqGRCh37_exon_slop_bed/2019-11-13_RefSeqGRCh37_slop145.bed\n",
      "\n",
      "\n",
      "Success! File saved:\n",
      "/Users/pbousounis/Experiments/2019-10-29_hg19mod/RefSeq-ClinVar_GRCh37_slop_region_recovery/data/2019-11-13_RefSeqGRCh37_exon_slop_bed/2019-11-13_RefSeqGRCh37_slop150.bed\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# for each N in {0, 5, 10, .., 155} (non-inclusive) run bedtools slop adding N bases to each end of each region\n",
    "for i in list(range(5, 155, 5)):\n",
    "    \n",
    "    # create the output filename\n",
    "    file_out = os.path.join(os.getcwd(), slop_dir_out, '{}_RefSeqGRCh37_slop{}.bed'.format(today, i))\n",
    "    \n",
    "    # perform slop: add 'i' bases to each end of each region\n",
    "    refseq_bedtool.slop(g=hg19_genome, b=i).saveas(file_out)\n",
    "    \n",
    "    # check file save\n",
    "    if os.path.isfile(os.path.join(os.getcwd(), file_out)):\n",
    "        print('\\nSuccess! File saved:\\n{}\\n'.format(file_out))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Get the ClinVar variant overlap counts for each RefSeq exonic region"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 524,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Success!\n",
      "File saved to: \n",
      "/Users/pbousounis/Experiments/2019-10-29_hg19mod/RefSeq-ClinVar_GRCh37_slop_region_recovery/data/../output/2019-11-14_RefSeqGRCh37_ClinVar_slop-intersect_bed/2019-11-13_RefSeqGRCh37_slop45_IN_2019-11-07-GRCh37_path-likely_path_headless.bed.\n",
      "\n",
      "Success!\n",
      "File saved to: \n",
      "/Users/pbousounis/Experiments/2019-10-29_hg19mod/RefSeq-ClinVar_GRCh37_slop_region_recovery/data/../output/2019-11-14_RefSeqGRCh37_ClinVar_slop-intersect_bed/2019-11-13_RefSeqGRCh37_slop50_IN_2019-11-07-GRCh37_path-likely_path_headless.bed.\n",
      "\n",
      "Success!\n",
      "File saved to: \n",
      "/Users/pbousounis/Experiments/2019-10-29_hg19mod/RefSeq-ClinVar_GRCh37_slop_region_recovery/data/../output/2019-11-14_RefSeqGRCh37_ClinVar_slop-intersect_bed/2019-11-13_RefSeqGRCh37_slop85_IN_2019-11-07-GRCh37_path-likely_path_headless.bed.\n",
      "\n",
      "Success!\n",
      "File saved to: \n",
      "/Users/pbousounis/Experiments/2019-10-29_hg19mod/RefSeq-ClinVar_GRCh37_slop_region_recovery/data/../output/2019-11-14_RefSeqGRCh37_ClinVar_slop-intersect_bed/2019-11-13_RefSeqGRCh37_slop90_IN_2019-11-07-GRCh37_path-likely_path_headless.bed.\n",
      "\n",
      "Success!\n",
      "File saved to: \n",
      "/Users/pbousounis/Experiments/2019-10-29_hg19mod/RefSeq-ClinVar_GRCh37_slop_region_recovery/data/../output/2019-11-14_RefSeqGRCh37_ClinVar_slop-intersect_bed/2019-11-13_RefSeqGRCh37_slop80_IN_2019-11-07-GRCh37_path-likely_path_headless.bed.\n",
      "\n",
      "Success!\n",
      "File saved to: \n",
      "/Users/pbousounis/Experiments/2019-10-29_hg19mod/RefSeq-ClinVar_GRCh37_slop_region_recovery/data/../output/2019-11-14_RefSeqGRCh37_ClinVar_slop-intersect_bed/2019-11-13_RefSeqGRCh37_slop95_IN_2019-11-07-GRCh37_path-likely_path_headless.bed.\n",
      "\n",
      "Success!\n",
      "File saved to: \n",
      "/Users/pbousounis/Experiments/2019-10-29_hg19mod/RefSeq-ClinVar_GRCh37_slop_region_recovery/data/../output/2019-11-14_RefSeqGRCh37_ClinVar_slop-intersect_bed/2019-11-13_RefSeqGRCh37_slop40_IN_2019-11-07-GRCh37_path-likely_path_headless.bed.\n",
      "\n",
      "Success!\n",
      "File saved to: \n",
      "/Users/pbousounis/Experiments/2019-10-29_hg19mod/RefSeq-ClinVar_GRCh37_slop_region_recovery/data/../output/2019-11-14_RefSeqGRCh37_ClinVar_slop-intersect_bed/2019-11-13_RefSeqGRCh37_slop55_IN_2019-11-07-GRCh37_path-likely_path_headless.bed.\n",
      "\n",
      "Success!\n",
      "File saved to: \n",
      "/Users/pbousounis/Experiments/2019-10-29_hg19mod/RefSeq-ClinVar_GRCh37_slop_region_recovery/data/../output/2019-11-14_RefSeqGRCh37_ClinVar_slop-intersect_bed/2019-11-13_RefSeqGRCh37_slop100_IN_2019-11-07-GRCh37_path-likely_path_headless.bed.\n",
      "\n",
      "Success!\n",
      "File saved to: \n",
      "/Users/pbousounis/Experiments/2019-10-29_hg19mod/RefSeq-ClinVar_GRCh37_slop_region_recovery/data/../output/2019-11-14_RefSeqGRCh37_ClinVar_slop-intersect_bed/2019-11-13_RefSeqGRCh37_slop115_IN_2019-11-07-GRCh37_path-likely_path_headless.bed.\n",
      "\n",
      "Success!\n",
      "File saved to: \n",
      "/Users/pbousounis/Experiments/2019-10-29_hg19mod/RefSeq-ClinVar_GRCh37_slop_region_recovery/data/../output/2019-11-14_RefSeqGRCh37_ClinVar_slop-intersect_bed/2019-11-13_RefSeqGRCh37_slop25_IN_2019-11-07-GRCh37_path-likely_path_headless.bed.\n",
      "\n",
      "Success!\n",
      "File saved to: \n",
      "/Users/pbousounis/Experiments/2019-10-29_hg19mod/RefSeq-ClinVar_GRCh37_slop_region_recovery/data/../output/2019-11-14_RefSeqGRCh37_ClinVar_slop-intersect_bed/2019-11-13_RefSeqGRCh37_slop30_IN_2019-11-07-GRCh37_path-likely_path_headless.bed.\n",
      "\n",
      "Success!\n",
      "File saved to: \n",
      "/Users/pbousounis/Experiments/2019-10-29_hg19mod/RefSeq-ClinVar_GRCh37_slop_region_recovery/data/../output/2019-11-14_RefSeqGRCh37_ClinVar_slop-intersect_bed/2019-11-13_RefSeqGRCh37_slop20_IN_2019-11-07-GRCh37_path-likely_path_headless.bed.\n",
      "\n",
      "Success!\n",
      "File saved to: \n",
      "/Users/pbousounis/Experiments/2019-10-29_hg19mod/RefSeq-ClinVar_GRCh37_slop_region_recovery/data/../output/2019-11-14_RefSeqGRCh37_ClinVar_slop-intersect_bed/2019-11-13_RefSeqGRCh37_slop35_IN_2019-11-07-GRCh37_path-likely_path_headless.bed.\n",
      "\n",
      "Success!\n",
      "File saved to: \n",
      "/Users/pbousounis/Experiments/2019-10-29_hg19mod/RefSeq-ClinVar_GRCh37_slop_region_recovery/data/../output/2019-11-14_RefSeqGRCh37_ClinVar_slop-intersect_bed/2019-11-13_RefSeqGRCh37_slop105_IN_2019-11-07-GRCh37_path-likely_path_headless.bed.\n",
      "\n",
      "Success!\n",
      "File saved to: \n",
      "/Users/pbousounis/Experiments/2019-10-29_hg19mod/RefSeq-ClinVar_GRCh37_slop_region_recovery/data/../output/2019-11-14_RefSeqGRCh37_ClinVar_slop-intersect_bed/2019-11-13_RefSeqGRCh37_slop110_IN_2019-11-07-GRCh37_path-likely_path_headless.bed.\n",
      "\n",
      "Success!\n",
      "File saved to: \n",
      "/Users/pbousounis/Experiments/2019-10-29_hg19mod/RefSeq-ClinVar_GRCh37_slop_region_recovery/data/../output/2019-11-14_RefSeqGRCh37_ClinVar_slop-intersect_bed/2019-11-13_RefSeqGRCh37_slop135_IN_2019-11-07-GRCh37_path-likely_path_headless.bed.\n",
      "\n",
      "Success!\n",
      "File saved to: \n",
      "/Users/pbousounis/Experiments/2019-10-29_hg19mod/RefSeq-ClinVar_GRCh37_slop_region_recovery/data/../output/2019-11-14_RefSeqGRCh37_ClinVar_slop-intersect_bed/2019-11-13_RefSeqGRCh37_slop120_IN_2019-11-07-GRCh37_path-likely_path_headless.bed.\n",
      "\n",
      "Success!\n",
      "File saved to: \n",
      "/Users/pbousounis/Experiments/2019-10-29_hg19mod/RefSeq-ClinVar_GRCh37_slop_region_recovery/data/../output/2019-11-14_RefSeqGRCh37_ClinVar_slop-intersect_bed/2019-11-13_RefSeqGRCh37_slop10_IN_2019-11-07-GRCh37_path-likely_path_headless.bed.\n",
      "\n",
      "Success!\n",
      "File saved to: \n",
      "/Users/pbousounis/Experiments/2019-10-29_hg19mod/RefSeq-ClinVar_GRCh37_slop_region_recovery/data/../output/2019-11-14_RefSeqGRCh37_ClinVar_slop-intersect_bed/2019-11-13_RefSeqGRCh37_slop15_IN_2019-11-07-GRCh37_path-likely_path_headless.bed.\n",
      "\n",
      "Success!\n",
      "File saved to: \n",
      "/Users/pbousounis/Experiments/2019-10-29_hg19mod/RefSeq-ClinVar_GRCh37_slop_region_recovery/data/../output/2019-11-14_RefSeqGRCh37_ClinVar_slop-intersect_bed/2019-11-13_RefSeqGRCh37_slop130_IN_2019-11-07-GRCh37_path-likely_path_headless.bed.\n",
      "\n",
      "Success!\n",
      "File saved to: \n",
      "/Users/pbousounis/Experiments/2019-10-29_hg19mod/RefSeq-ClinVar_GRCh37_slop_region_recovery/data/../output/2019-11-14_RefSeqGRCh37_ClinVar_slop-intersect_bed/2019-11-13_RefSeqGRCh37_slop125_IN_2019-11-07-GRCh37_path-likely_path_headless.bed.\n",
      "\n",
      "Success!\n",
      "File saved to: \n",
      "/Users/pbousounis/Experiments/2019-10-29_hg19mod/RefSeq-ClinVar_GRCh37_slop_region_recovery/data/../output/2019-11-14_RefSeqGRCh37_ClinVar_slop-intersect_bed/2019-11-13_RefSeqGRCh37_slop70_IN_2019-11-07-GRCh37_path-likely_path_headless.bed.\n",
      "\n",
      "Success!\n",
      "File saved to: \n",
      "/Users/pbousounis/Experiments/2019-10-29_hg19mod/RefSeq-ClinVar_GRCh37_slop_region_recovery/data/../output/2019-11-14_RefSeqGRCh37_ClinVar_slop-intersect_bed/2019-11-13_RefSeqGRCh37_slop65_IN_2019-11-07-GRCh37_path-likely_path_headless.bed.\n",
      "\n",
      "Success!\n",
      "File saved to: \n",
      "/Users/pbousounis/Experiments/2019-10-29_hg19mod/RefSeq-ClinVar_GRCh37_slop_region_recovery/data/../output/2019-11-14_RefSeqGRCh37_ClinVar_slop-intersect_bed/2019-11-13_RefSeqGRCh37_slop140_IN_2019-11-07-GRCh37_path-likely_path_headless.bed.\n",
      "\n",
      "Success!\n",
      "File saved to: \n",
      "/Users/pbousounis/Experiments/2019-10-29_hg19mod/RefSeq-ClinVar_GRCh37_slop_region_recovery/data/../output/2019-11-14_RefSeqGRCh37_ClinVar_slop-intersect_bed/2019-11-13_RefSeqGRCh37_slop5_IN_2019-11-07-GRCh37_path-likely_path_headless.bed.\n",
      "\n",
      "Success!\n",
      "File saved to: \n",
      "/Users/pbousounis/Experiments/2019-10-29_hg19mod/RefSeq-ClinVar_GRCh37_slop_region_recovery/data/../output/2019-11-14_RefSeqGRCh37_ClinVar_slop-intersect_bed/2019-11-13_RefSeqGRCh37_slop150_IN_2019-11-07-GRCh37_path-likely_path_headless.bed.\n",
      "\n",
      "Success!\n",
      "File saved to: \n",
      "/Users/pbousounis/Experiments/2019-10-29_hg19mod/RefSeq-ClinVar_GRCh37_slop_region_recovery/data/../output/2019-11-14_RefSeqGRCh37_ClinVar_slop-intersect_bed/2019-11-13_RefSeqGRCh37_slop145_IN_2019-11-07-GRCh37_path-likely_path_headless.bed.\n",
      "\n",
      "Success!\n",
      "File saved to: \n",
      "/Users/pbousounis/Experiments/2019-10-29_hg19mod/RefSeq-ClinVar_GRCh37_slop_region_recovery/data/../output/2019-11-14_RefSeqGRCh37_ClinVar_slop-intersect_bed/2019-11-13_RefSeqGRCh37_slop75_IN_2019-11-07-GRCh37_path-likely_path_headless.bed.\n",
      "\n",
      "Success!\n",
      "File saved to: \n",
      "/Users/pbousounis/Experiments/2019-10-29_hg19mod/RefSeq-ClinVar_GRCh37_slop_region_recovery/data/../output/2019-11-14_RefSeqGRCh37_ClinVar_slop-intersect_bed/2019-11-13_RefSeqGRCh37_slop60_IN_2019-11-07-GRCh37_path-likely_path_headless.bed.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# the clinvar bed file to intersect\n",
    "bed_a_file = '2019-11-07-GRCh37_path-likely_path_headless.bed'\n",
    "\n",
    "# list of slopped refseq bed files to intersect\n",
    "slop_files = glob.glob('2019-11-13_RefSeqGRCh37_exon_slop_bed/*RefSeqGRCh37_slop*.bed')\n",
    "\n",
    "# create/specify output directory\n",
    "slop_int_out = os.path.join(('../output/' + today + \"_RefSeqGRCh37_ClinVar_slop-intersect_bed\"))\n",
    "pathlib.Path(slop_int_out).mkdir(exist_ok=True)\n",
    "\n",
    "# create empty dict \n",
    "for i in slop_files:\n",
    "    slop_num = re.findall(r'\\d+', i)[-1]\n",
    "    fname = '{}_RefSeqGRCh37_ClinVar_recovery_{}.bed'.format(today, slop_num)\n",
    "    bed_overlap(bed_a_file, i, slop_int_out)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Concatenate bed_overlap slop files "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 525,
   "metadata": {},
   "outputs": [],
   "source": [
    "# list overlapped slop files\n",
    "ovl_slop_files = glob.glob(slop_int_out + '/*RefSeqGRCh37_slop*IN*GRCh37_path-likely_path_headless.bed')\n",
    "\n",
    "# read in the files and concatenate\n",
    "bedcols = ['chr', 'start', 'end', 'name', 'exon_recovery']\n",
    "df = pd.DataFrame(columns=bedcols)\n",
    "for file in ovl_slop_files:\n",
    "    \n",
    "    # import file as dataframe\n",
    "    tmp = pd.read_csv(file, sep='\\t', header=None, names=bedcols)\n",
    "    \n",
    "    # extract transcript IDs and gene names to their own columns\n",
    "    tmp['tx_id'] = tmp['name'].str.extract(r'(\\w+-)(N(M|R)_\\d+)')[1]\n",
    "    tmp['gene'] = tmp['name'].str.split('-').str[0]\n",
    "    \n",
    "    # add column with slop file ID\n",
    "    tmp['slop_len'] = re.findall(r'slop\\d+', file)[0]\n",
    "    \n",
    "    # Filter: remove rows (regions) with no ClinVar overlap\n",
    "    tmp = tmp[tmp.exon_recovery != 0]\n",
    "    \n",
    "    # append to df\n",
    "    df = df.append(tmp, sort=False)\n",
    "    del(tmp)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 331,
   "metadata": {},
   "outputs": [],
   "source": [
    "# get transcript recovery by slop distance\n",
    "df['tx_slop_recovery'] = df['exon_recovery'].groupby([df['tx_id'], df['slop_len']]).transform('sum')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# get transcript recovery by slop distance ##CHECK LABEL exon_recovery\n",
    "df['tx_slop_recovery'] = df['exon_recovery'].groupby([df['tx_id'], df['slop_len']]).transform('sum')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 334,
   "metadata": {},
   "outputs": [],
   "source": [
    "# spread df by slop length\n",
    "slop_df = df.pivot_table(index='tx_id', columns='slop_len', values='tx_slop_recovery')\n",
    "slop_df = slop_df[natsorted(slop_df.columns)].add_suffix('_recovery')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# spread df by slop length\n",
    "slop_df = df.pivot_table(index='tx_id', columns='slop_len', values='tx_slop_recovery')\n",
    "slop_df = slop_df[natsorted(slop_df.columns)].add_suffix('_recovery')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 337,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "<class 'pandas.core.frame.DataFrame'>\n",
      "Index: 14197 entries, NM_000016 to NR_164148\n",
      "Data columns (total 30 columns):\n",
      "slop5_recovery      13937 non-null float64\n",
      "slop10_recovery     13963 non-null float64\n",
      "slop15_recovery     13977 non-null float64\n",
      "slop20_recovery     13998 non-null float64\n",
      "slop25_recovery     14006 non-null float64\n",
      "slop30_recovery     14018 non-null float64\n",
      "slop35_recovery     14028 non-null float64\n",
      "slop40_recovery     14033 non-null float64\n",
      "slop45_recovery     14036 non-null float64\n",
      "slop50_recovery     14040 non-null float64\n",
      "slop55_recovery     14048 non-null float64\n",
      "slop60_recovery     14055 non-null float64\n",
      "slop65_recovery     14064 non-null float64\n",
      "slop70_recovery     14066 non-null float64\n",
      "slop75_recovery     14067 non-null float64\n",
      "slop80_recovery     14087 non-null float64\n",
      "slop85_recovery     14097 non-null float64\n",
      "slop90_recovery     14098 non-null float64\n",
      "slop95_recovery     14107 non-null float64\n",
      "slop100_recovery    14117 non-null float64\n",
      "slop105_recovery    14124 non-null float64\n",
      "slop110_recovery    14125 non-null float64\n",
      "slop115_recovery    14139 non-null float64\n",
      "slop120_recovery    14147 non-null float64\n",
      "slop125_recovery    14150 non-null float64\n",
      "slop130_recovery    14156 non-null float64\n",
      "slop135_recovery    14158 non-null float64\n",
      "slop140_recovery    14174 non-null float64\n",
      "slop145_recovery    14193 non-null float64\n",
      "slop150_recovery    14197 non-null float64\n",
      "dtypes: float64(30)\n",
      "memory usage: 3.4+ MB\n",
      "\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>slop_len</th>\n",
       "      <th>slop5_recovery</th>\n",
       "      <th>slop10_recovery</th>\n",
       "      <th>slop15_recovery</th>\n",
       "      <th>slop20_recovery</th>\n",
       "      <th>slop25_recovery</th>\n",
       "      <th>slop30_recovery</th>\n",
       "      <th>slop35_recovery</th>\n",
       "      <th>slop40_recovery</th>\n",
       "      <th>slop45_recovery</th>\n",
       "      <th>slop50_recovery</th>\n",
       "      <th>...</th>\n",
       "      <th>slop105_recovery</th>\n",
       "      <th>slop110_recovery</th>\n",
       "      <th>slop115_recovery</th>\n",
       "      <th>slop120_recovery</th>\n",
       "      <th>slop125_recovery</th>\n",
       "      <th>slop130_recovery</th>\n",
       "      <th>slop135_recovery</th>\n",
       "      <th>slop140_recovery</th>\n",
       "      <th>slop145_recovery</th>\n",
       "      <th>slop150_recovery</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>tx_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>NM_000016</th>\n",
       "      <td>73.0</td>\n",
       "      <td>74.0</td>\n",
       "      <td>74.0</td>\n",
       "      <td>74.0</td>\n",
       "      <td>74.0</td>\n",
       "      <td>74.0</td>\n",
       "      <td>74.0</td>\n",
       "      <td>74.0</td>\n",
       "      <td>74.0</td>\n",
       "      <td>74.0</td>\n",
       "      <td>...</td>\n",
       "      <td>74.0</td>\n",
       "      <td>74.0</td>\n",
       "      <td>78.0</td>\n",
       "      <td>79.0</td>\n",
       "      <td>79.0</td>\n",
       "      <td>79.0</td>\n",
       "      <td>79.0</td>\n",
       "      <td>79.0</td>\n",
       "      <td>79.0</td>\n",
       "      <td>80.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NM_000017</th>\n",
       "      <td>30.0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>...</td>\n",
       "      <td>35.0</td>\n",
       "      <td>35.0</td>\n",
       "      <td>35.0</td>\n",
       "      <td>35.0</td>\n",
       "      <td>35.0</td>\n",
       "      <td>35.0</td>\n",
       "      <td>35.0</td>\n",
       "      <td>35.0</td>\n",
       "      <td>35.0</td>\n",
       "      <td>35.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NM_000018</th>\n",
       "      <td>97.0</td>\n",
       "      <td>97.0</td>\n",
       "      <td>97.0</td>\n",
       "      <td>97.0</td>\n",
       "      <td>97.0</td>\n",
       "      <td>97.0</td>\n",
       "      <td>97.0</td>\n",
       "      <td>97.0</td>\n",
       "      <td>97.0</td>\n",
       "      <td>97.0</td>\n",
       "      <td>...</td>\n",
       "      <td>129.0</td>\n",
       "      <td>136.0</td>\n",
       "      <td>139.0</td>\n",
       "      <td>146.0</td>\n",
       "      <td>150.0</td>\n",
       "      <td>155.0</td>\n",
       "      <td>158.0</td>\n",
       "      <td>165.0</td>\n",
       "      <td>171.0</td>\n",
       "      <td>178.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NM_000019</th>\n",
       "      <td>28.0</td>\n",
       "      <td>29.0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>...</td>\n",
       "      <td>31.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>31.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NM_000020</th>\n",
       "      <td>129.0</td>\n",
       "      <td>129.0</td>\n",
       "      <td>129.0</td>\n",
       "      <td>129.0</td>\n",
       "      <td>129.0</td>\n",
       "      <td>129.0</td>\n",
       "      <td>129.0</td>\n",
       "      <td>129.0</td>\n",
       "      <td>129.0</td>\n",
       "      <td>129.0</td>\n",
       "      <td>...</td>\n",
       "      <td>129.0</td>\n",
       "      <td>129.0</td>\n",
       "      <td>129.0</td>\n",
       "      <td>129.0</td>\n",
       "      <td>129.0</td>\n",
       "      <td>129.0</td>\n",
       "      <td>129.0</td>\n",
       "      <td>129.0</td>\n",
       "      <td>129.0</td>\n",
       "      <td>129.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 30 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "slop_len   slop5_recovery  slop10_recovery  slop15_recovery  slop20_recovery  \\\n",
       "tx_id                                                                          \n",
       "NM_000016            73.0             74.0             74.0             74.0   \n",
       "NM_000017            30.0             30.0             30.0             30.0   \n",
       "NM_000018            97.0             97.0             97.0             97.0   \n",
       "NM_000019            28.0             29.0             30.0             30.0   \n",
       "NM_000020           129.0            129.0            129.0            129.0   \n",
       "\n",
       "slop_len   slop25_recovery  slop30_recovery  slop35_recovery  slop40_recovery  \\\n",
       "tx_id                                                                           \n",
       "NM_000016             74.0             74.0             74.0             74.0   \n",
       "NM_000017             30.0             30.0             30.0             30.0   \n",
       "NM_000018             97.0             97.0             97.0             97.0   \n",
       "NM_000019             30.0             30.0             30.0             30.0   \n",
       "NM_000020            129.0            129.0            129.0            129.0   \n",
       "\n",
       "slop_len   slop45_recovery  slop50_recovery  ...  slop105_recovery  \\\n",
       "tx_id                                        ...                     \n",
       "NM_000016             74.0             74.0  ...              74.0   \n",
       "NM_000017             30.0             30.0  ...              35.0   \n",
       "NM_000018             97.0             97.0  ...             129.0   \n",
       "NM_000019             30.0             30.0  ...              31.0   \n",
       "NM_000020            129.0            129.0  ...             129.0   \n",
       "\n",
       "slop_len   slop110_recovery  slop115_recovery  slop120_recovery  \\\n",
       "tx_id                                                             \n",
       "NM_000016              74.0              78.0              79.0   \n",
       "NM_000017              35.0              35.0              35.0   \n",
       "NM_000018             136.0             139.0             146.0   \n",
       "NM_000019              31.0              31.0              31.0   \n",
       "NM_000020             129.0             129.0             129.0   \n",
       "\n",
       "slop_len   slop125_recovery  slop130_recovery  slop135_recovery  \\\n",
       "tx_id                                                             \n",
       "NM_000016              79.0              79.0              79.0   \n",
       "NM_000017              35.0              35.0              35.0   \n",
       "NM_000018             150.0             155.0             158.0   \n",
       "NM_000019              31.0              31.0              31.0   \n",
       "NM_000020             129.0             129.0             129.0   \n",
       "\n",
       "slop_len   slop140_recovery  slop145_recovery  slop150_recovery  \n",
       "tx_id                                                            \n",
       "NM_000016              79.0              79.0              80.0  \n",
       "NM_000017              35.0              35.0              35.0  \n",
       "NM_000018             165.0             171.0             178.0  \n",
       "NM_000019              31.0              31.0              31.0  \n",
       "NM_000020             129.0             129.0             129.0  \n",
       "\n",
       "[5 rows x 30 columns]"
      ]
     },
     "execution_count": 337,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print('\\n')\n",
    "slop_df.info()\n",
    "print('\\n')\n",
    "slop_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 535,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tx_id</th>\n",
       "      <th>gene</th>\n",
       "      <th>slop0_recovery</th>\n",
       "      <th>gene_recovery</th>\n",
       "      <th>slop0_recovery%</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NM_005101</td>\n",
       "      <td>ISG15</td>\n",
       "      <td>3</td>\n",
       "      <td>3.0</td>\n",
       "      <td>100.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NM_001305275</td>\n",
       "      <td>AGRN</td>\n",
       "      <td>15</td>\n",
       "      <td>15.0</td>\n",
       "      <td>100.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NM_198576</td>\n",
       "      <td>AGRN</td>\n",
       "      <td>15</td>\n",
       "      <td>15.0</td>\n",
       "      <td>100.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NM_001364727</td>\n",
       "      <td>AGRN</td>\n",
       "      <td>13</td>\n",
       "      <td>15.0</td>\n",
       "      <td>86.666667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NM_080605</td>\n",
       "      <td>B3GALT6</td>\n",
       "      <td>9</td>\n",
       "      <td>9.0</td>\n",
       "      <td>100.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          tx_id     gene  slop0_recovery  gene_recovery  slop0_recovery%\n",
       "0     NM_005101    ISG15               3            3.0       100.000000\n",
       "1  NM_001305275     AGRN              15           15.0       100.000000\n",
       "2     NM_198576     AGRN              15           15.0       100.000000\n",
       "3  NM_001364727     AGRN              13           15.0        86.666667\n",
       "4     NM_080605  B3GALT6               9            9.0       100.000000"
      ]
     },
     "execution_count": 535,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# import the transcript percent recovery table\n",
    "tx_rec_file = '2019-11-12_RefSeq-ClinVar_transcript-recovery.tsv'\n",
    "tx_rec_cols = ['tx_id', 'gene', 'slop0_recovery', 'gene_recovery', 'slop0_recovery%']\n",
    "tx_rec = pd.read_csv(tx_rec_file, sep='\\t', skiprows=1, names=tx_rec_cols)\n",
    "tx_rec.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 351,
   "metadata": {},
   "outputs": [],
   "source": [
    "# add gene_recovery column to slop_df\n",
    "slop_tx = slop_df.merge(tx_rec[['tx_id', 'gene_recovery']], how='left', on='tx_id')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# add gene_recovery column to slop_df\n",
    "slop_tx = slop_df.merge(tx_rec[['tx_id', 'gene_recovery']], how='left', on='tx_id')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 352,
   "metadata": {},
   "outputs": [],
   "source": [
    "# calculate percent recovery for each slop length\n",
    "for col in slop_tx.loc[:, slop_tx.columns.str.contains('slop')].columns:\n",
    "    new_col = col.split('_')[0] + '_recovery%'\n",
    "    slop_tx[new_col] = slop_tx[col] / slop_tx['gene_recovery'] * 100"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 365,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tx_id</th>\n",
       "      <th>gene</th>\n",
       "      <th>slop0_recovery</th>\n",
       "      <th>slop0_recovery%</th>\n",
       "      <th>slop5_recovery</th>\n",
       "      <th>slop10_recovery</th>\n",
       "      <th>slop15_recovery</th>\n",
       "      <th>slop20_recovery</th>\n",
       "      <th>slop25_recovery</th>\n",
       "      <th>slop30_recovery</th>\n",
       "      <th>...</th>\n",
       "      <th>slop105_recovery%</th>\n",
       "      <th>slop110_recovery%</th>\n",
       "      <th>slop115_recovery%</th>\n",
       "      <th>slop120_recovery%</th>\n",
       "      <th>slop125_recovery%</th>\n",
       "      <th>slop130_recovery%</th>\n",
       "      <th>slop135_recovery%</th>\n",
       "      <th>slop140_recovery%</th>\n",
       "      <th>slop145_recovery%</th>\n",
       "      <th>slop150_recovery%</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NM_005101</td>\n",
       "      <td>ISG15</td>\n",
       "      <td>3</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>...</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>100.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NM_001305275</td>\n",
       "      <td>AGRN</td>\n",
       "      <td>15</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>15.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>...</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>106.666667</td>\n",
       "      <td>113.333333</td>\n",
       "      <td>113.333333</td>\n",
       "      <td>113.333333</td>\n",
       "      <td>133.333333</td>\n",
       "      <td>133.333333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NM_198576</td>\n",
       "      <td>AGRN</td>\n",
       "      <td>15</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>15.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>...</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>106.666667</td>\n",
       "      <td>113.333333</td>\n",
       "      <td>113.333333</td>\n",
       "      <td>113.333333</td>\n",
       "      <td>133.333333</td>\n",
       "      <td>133.333333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NM_001364727</td>\n",
       "      <td>AGRN</td>\n",
       "      <td>13</td>\n",
       "      <td>86.666667</td>\n",
       "      <td>13.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>...</td>\n",
       "      <td>86.666667</td>\n",
       "      <td>86.666667</td>\n",
       "      <td>86.666667</td>\n",
       "      <td>86.666667</td>\n",
       "      <td>93.333333</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>120.000000</td>\n",
       "      <td>120.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NM_080605</td>\n",
       "      <td>B3GALT6</td>\n",
       "      <td>9</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>9.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>...</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>100.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 65 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          tx_id     gene  slop0_recovery  slop0_recovery%  slop5_recovery  \\\n",
       "0     NM_005101    ISG15               3       100.000000             3.0   \n",
       "1  NM_001305275     AGRN              15       100.000000            15.0   \n",
       "2     NM_198576     AGRN              15       100.000000            15.0   \n",
       "3  NM_001364727     AGRN              13        86.666667            13.0   \n",
       "4     NM_080605  B3GALT6               9       100.000000             9.0   \n",
       "\n",
       "   slop10_recovery  slop15_recovery  slop20_recovery  slop25_recovery  \\\n",
       "0              3.0              3.0              3.0              3.0   \n",
       "1             15.0             15.0             15.0             15.0   \n",
       "2             15.0             15.0             15.0             15.0   \n",
       "3             13.0             13.0             13.0             13.0   \n",
       "4              9.0              9.0              9.0              9.0   \n",
       "\n",
       "   slop30_recovery  ...  slop105_recovery%  slop110_recovery%  \\\n",
       "0              3.0  ...         100.000000         100.000000   \n",
       "1             15.0  ...         100.000000         100.000000   \n",
       "2             15.0  ...         100.000000         100.000000   \n",
       "3             13.0  ...          86.666667          86.666667   \n",
       "4              9.0  ...         100.000000         100.000000   \n",
       "\n",
       "   slop115_recovery%  slop120_recovery%  slop125_recovery%  slop130_recovery%  \\\n",
       "0         100.000000         100.000000         100.000000         100.000000   \n",
       "1         100.000000         100.000000         106.666667         113.333333   \n",
       "2         100.000000         100.000000         106.666667         113.333333   \n",
       "3          86.666667          86.666667          93.333333         100.000000   \n",
       "4         100.000000         100.000000         100.000000         100.000000   \n",
       "\n",
       "   slop135_recovery%  slop140_recovery%  slop145_recovery%  slop150_recovery%  \n",
       "0         100.000000         100.000000         100.000000         100.000000  \n",
       "1         113.333333         113.333333         133.333333         133.333333  \n",
       "2         113.333333         113.333333         133.333333         133.333333  \n",
       "3         100.000000         100.000000         120.000000         120.000000  \n",
       "4         100.000000         100.000000         100.000000         100.000000  \n",
       "\n",
       "[5 rows x 65 columns]"
      ]
     },
     "execution_count": 365,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tmp_slop_table = pd.merge(tx_rec[['tx_id', 'gene', 'slop0_recovery', 'slop0_recovery%']], slop_tx, how='left', on='tx_id')\n",
    "tmp_slop_table.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 434,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tx_id</th>\n",
       "      <th>gene</th>\n",
       "      <th>slop0_recovery</th>\n",
       "      <th>slop0_recovery%</th>\n",
       "      <th>slop5_recovery</th>\n",
       "      <th>slop5_recovery%</th>\n",
       "      <th>slop10_recovery</th>\n",
       "      <th>slop10_recovery%</th>\n",
       "      <th>slop15_recovery</th>\n",
       "      <th>slop15_recovery%</th>\n",
       "      <th>...</th>\n",
       "      <th>slop130_recovery</th>\n",
       "      <th>slop130_recovery%</th>\n",
       "      <th>slop135_recovery</th>\n",
       "      <th>slop135_recovery%</th>\n",
       "      <th>slop140_recovery</th>\n",
       "      <th>slop140_recovery%</th>\n",
       "      <th>slop145_recovery</th>\n",
       "      <th>slop145_recovery%</th>\n",
       "      <th>slop150_recovery</th>\n",
       "      <th>slop150_recovery%</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NM_005101</td>\n",
       "      <td>ISG15</td>\n",
       "      <td>3</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>3.0</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>3.0</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>3.0</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>3.0</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>3.0</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>3.0</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>3.0</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>3.0</td>\n",
       "      <td>100.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NM_001305275</td>\n",
       "      <td>AGRN</td>\n",
       "      <td>15</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>15.0</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>15.0</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>15.0</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>17.0</td>\n",
       "      <td>113.333333</td>\n",
       "      <td>17.0</td>\n",
       "      <td>113.333333</td>\n",
       "      <td>17.0</td>\n",
       "      <td>113.333333</td>\n",
       "      <td>20.0</td>\n",
       "      <td>133.333333</td>\n",
       "      <td>20.0</td>\n",
       "      <td>133.333333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NM_198576</td>\n",
       "      <td>AGRN</td>\n",
       "      <td>15</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>15.0</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>15.0</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>15.0</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>17.0</td>\n",
       "      <td>113.333333</td>\n",
       "      <td>17.0</td>\n",
       "      <td>113.333333</td>\n",
       "      <td>17.0</td>\n",
       "      <td>113.333333</td>\n",
       "      <td>20.0</td>\n",
       "      <td>133.333333</td>\n",
       "      <td>20.0</td>\n",
       "      <td>133.333333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NM_001364727</td>\n",
       "      <td>AGRN</td>\n",
       "      <td>13</td>\n",
       "      <td>86.666667</td>\n",
       "      <td>13.0</td>\n",
       "      <td>86.666667</td>\n",
       "      <td>13.0</td>\n",
       "      <td>86.666667</td>\n",
       "      <td>13.0</td>\n",
       "      <td>86.666667</td>\n",
       "      <td>...</td>\n",
       "      <td>15.0</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>15.0</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>15.0</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>18.0</td>\n",
       "      <td>120.000000</td>\n",
       "      <td>18.0</td>\n",
       "      <td>120.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NM_080605</td>\n",
       "      <td>B3GALT6</td>\n",
       "      <td>9</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>9.0</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>9.0</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>9.0</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>9.0</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>9.0</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>9.0</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>9.0</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>9.0</td>\n",
       "      <td>100.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 64 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          tx_id     gene  slop0_recovery  slop0_recovery%  slop5_recovery  \\\n",
       "0     NM_005101    ISG15               3       100.000000             3.0   \n",
       "1  NM_001305275     AGRN              15       100.000000            15.0   \n",
       "2     NM_198576     AGRN              15       100.000000            15.0   \n",
       "3  NM_001364727     AGRN              13        86.666667            13.0   \n",
       "4     NM_080605  B3GALT6               9       100.000000             9.0   \n",
       "\n",
       "   slop5_recovery%  slop10_recovery  slop10_recovery%  slop15_recovery  \\\n",
       "0       100.000000              3.0        100.000000              3.0   \n",
       "1       100.000000             15.0        100.000000             15.0   \n",
       "2       100.000000             15.0        100.000000             15.0   \n",
       "3        86.666667             13.0         86.666667             13.0   \n",
       "4       100.000000              9.0        100.000000              9.0   \n",
       "\n",
       "   slop15_recovery%  ...  slop130_recovery  slop130_recovery%  \\\n",
       "0        100.000000  ...               3.0         100.000000   \n",
       "1        100.000000  ...              17.0         113.333333   \n",
       "2        100.000000  ...              17.0         113.333333   \n",
       "3         86.666667  ...              15.0         100.000000   \n",
       "4        100.000000  ...               9.0         100.000000   \n",
       "\n",
       "   slop135_recovery  slop135_recovery%  slop140_recovery  slop140_recovery%  \\\n",
       "0               3.0         100.000000               3.0         100.000000   \n",
       "1              17.0         113.333333              17.0         113.333333   \n",
       "2              17.0         113.333333              17.0         113.333333   \n",
       "3              15.0         100.000000              15.0         100.000000   \n",
       "4               9.0         100.000000               9.0         100.000000   \n",
       "\n",
       "   slop145_recovery  slop145_recovery%  slop150_recovery  slop150_recovery%  \n",
       "0               3.0         100.000000               3.0         100.000000  \n",
       "1              20.0         133.333333              20.0         133.333333  \n",
       "2              20.0         133.333333              20.0         133.333333  \n",
       "3              18.0         120.000000              18.0         120.000000  \n",
       "4               9.0         100.000000               9.0         100.000000  \n",
       "\n",
       "[5 rows x 64 columns]"
      ]
     },
     "execution_count": 434,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# reorder columns\n",
    "slop_cols_list = ['tx_id', 'gene'] + natsorted(tmp_slop_table.loc[:, tmp_slop_table.columns.str.contains('slop')])\n",
    "slop_table = tmp_slop_table[slop_cols_list]\n",
    "slop_table.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 460,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/anaconda3/envs/biopython/lib/python3.7/site-packages/pandas/core/frame.py:3509: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  self[k1] = value[k2]\n"
     ]
    }
   ],
   "source": [
    "# round _recovery% columns to two decimal places\n",
    "slop_table[natsorted(tmp_slop_table.loc[:, tmp_slop_table.columns.str.contains('_recovery%')])] = slop_table.loc[:, slop_table.columns.str.contains('_recovery%')].round(2)\n",
    "\n",
    "# convert recovery columns to integers\n",
    "slop_table[natsorted(tmp_slop_table.loc[:, tmp_slop_table.columns.str.contains('slop.+_recovery$')])] = slop_table[natsorted(tmp_slop_table.loc[:, tmp_slop_table.columns.str.contains('slop.+_recovery$')])].astype('int')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Create new column of top slop distances (<= 100% recovery)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 493,
   "metadata": {},
   "outputs": [],
   "source": [
    "# subset columns\n",
    "slop_tmp = slop_table.set_index('tx_id')\n",
    "slop_perc_cols = slop_tmp.loc[:, slop_tmp.columns.str.contains('_recovery%')]\n",
    "\n",
    "slop_perc_cols = slop_perc_cols[slop_perc_cols <= 100]\n",
    "\n",
    "# create the top slop column (= slop distance with maximum recovery <= 100%)\n",
    "# returns first (lowest) slop distance meeting above criteria\n",
    "slop_perc_cols['top_slop'] = slop_perc_cols.idxmax(axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 494,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tx_id</th>\n",
       "      <th>gene</th>\n",
       "      <th>gene_recovery</th>\n",
       "      <th>slop0_recovery</th>\n",
       "      <th>slop5_recovery</th>\n",
       "      <th>slop10_recovery</th>\n",
       "      <th>slop15_recovery</th>\n",
       "      <th>slop20_recovery</th>\n",
       "      <th>slop25_recovery</th>\n",
       "      <th>slop30_recovery</th>\n",
       "      <th>...</th>\n",
       "      <th>slop110_recovery%</th>\n",
       "      <th>slop115_recovery%</th>\n",
       "      <th>slop120_recovery%</th>\n",
       "      <th>slop125_recovery%</th>\n",
       "      <th>slop130_recovery%</th>\n",
       "      <th>slop135_recovery%</th>\n",
       "      <th>slop140_recovery%</th>\n",
       "      <th>slop145_recovery%</th>\n",
       "      <th>slop150_recovery%</th>\n",
       "      <th>top_slop</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NM_005101</td>\n",
       "      <td>ISG15</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>...</td>\n",
       "      <td>100.00</td>\n",
       "      <td>100.00</td>\n",
       "      <td>100.00</td>\n",
       "      <td>100.00</td>\n",
       "      <td>100.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>slop0_recovery%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NM_001305275</td>\n",
       "      <td>AGRN</td>\n",
       "      <td>15.0</td>\n",
       "      <td>15</td>\n",
       "      <td>15</td>\n",
       "      <td>15</td>\n",
       "      <td>15</td>\n",
       "      <td>15</td>\n",
       "      <td>15</td>\n",
       "      <td>15</td>\n",
       "      <td>...</td>\n",
       "      <td>100.00</td>\n",
       "      <td>100.00</td>\n",
       "      <td>100.00</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>slop0_recovery%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NM_198576</td>\n",
       "      <td>AGRN</td>\n",
       "      <td>15.0</td>\n",
       "      <td>15</td>\n",
       "      <td>15</td>\n",
       "      <td>15</td>\n",
       "      <td>15</td>\n",
       "      <td>15</td>\n",
       "      <td>15</td>\n",
       "      <td>15</td>\n",
       "      <td>...</td>\n",
       "      <td>100.00</td>\n",
       "      <td>100.00</td>\n",
       "      <td>100.00</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>slop0_recovery%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NM_001364727</td>\n",
       "      <td>AGRN</td>\n",
       "      <td>15.0</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>...</td>\n",
       "      <td>86.67</td>\n",
       "      <td>86.67</td>\n",
       "      <td>86.67</td>\n",
       "      <td>93.33</td>\n",
       "      <td>100.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>slop130_recovery%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NM_080605</td>\n",
       "      <td>B3GALT6</td>\n",
       "      <td>9.0</td>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "      <td>...</td>\n",
       "      <td>100.00</td>\n",
       "      <td>100.00</td>\n",
       "      <td>100.00</td>\n",
       "      <td>100.00</td>\n",
       "      <td>100.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>slop0_recovery%</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 66 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          tx_id     gene  gene_recovery  slop0_recovery  slop5_recovery  \\\n",
       "0     NM_005101    ISG15            3.0               3               3   \n",
       "1  NM_001305275     AGRN           15.0              15              15   \n",
       "2     NM_198576     AGRN           15.0              15              15   \n",
       "3  NM_001364727     AGRN           15.0              13              13   \n",
       "4     NM_080605  B3GALT6            9.0               9               9   \n",
       "\n",
       "   slop10_recovery  slop15_recovery  slop20_recovery  slop25_recovery  \\\n",
       "0                3                3                3                3   \n",
       "1               15               15               15               15   \n",
       "2               15               15               15               15   \n",
       "3               13               13               13               13   \n",
       "4                9                9                9                9   \n",
       "\n",
       "   slop30_recovery  ...  slop110_recovery%  slop115_recovery%  \\\n",
       "0                3  ...             100.00             100.00   \n",
       "1               15  ...             100.00             100.00   \n",
       "2               15  ...             100.00             100.00   \n",
       "3               13  ...              86.67              86.67   \n",
       "4                9  ...             100.00             100.00   \n",
       "\n",
       "   slop120_recovery%  slop125_recovery%  slop130_recovery%  slop135_recovery%  \\\n",
       "0             100.00             100.00              100.0              100.0   \n",
       "1             100.00                NaN                NaN                NaN   \n",
       "2             100.00                NaN                NaN                NaN   \n",
       "3              86.67              93.33              100.0              100.0   \n",
       "4             100.00             100.00              100.0              100.0   \n",
       "\n",
       "   slop140_recovery%  slop145_recovery%  slop150_recovery%           top_slop  \n",
       "0              100.0              100.0              100.0    slop0_recovery%  \n",
       "1                NaN                NaN                NaN    slop0_recovery%  \n",
       "2                NaN                NaN                NaN    slop0_recovery%  \n",
       "3              100.0                NaN                NaN  slop130_recovery%  \n",
       "4              100.0              100.0              100.0    slop0_recovery%  \n",
       "\n",
       "[5 rows x 66 columns]"
      ]
     },
     "execution_count": 494,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# reconstitute the full slop table\n",
    "# trim _recovery% columns from slop_table \n",
    "slop_rec = slop_table.loc[:, slop_table.columns.str.contains('_recovery$') | slop_table.columns.str.contains('tx_id') | slop_table.columns.str.contains('gene')]\n",
    "\n",
    "# add top_slop column\n",
    "slop_table_top = slop_rec.merge(slop_perc_cols, on='tx_id', how='left', suffixes=(False, False))\n",
    "\n",
    "# add gene_recovery column\n",
    "gene_recovery_df = tx_rec.iloc[:, [0,3]]\n",
    "slop_master = slop_table_top.merge(gene_recovery_df, on='tx_id', how='left', suffixes=(False, False))\n",
    "\n",
    "# reorder columns\n",
    "slop_master_cols = slop_master.columns.tolist()\n",
    "\n",
    "slop_cols = ['tx_id', 'gene', 'top_slop', 'gene_recovery'] + natsorted(slop_master.loc[:, slop_master.columns.str.contains('slop.+recovery')])\n",
    "slop_master_table = slop_master[slop_cols]\n",
    "\n",
    "slop_master_table.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Save slop recovery table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 513,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Success! File saved to:\n",
      "../output/2019-11-13_RefSeq-ClinVar_GRCh37_slop-recovery-master.tsv\n"
     ]
    }
   ],
   "source": [
    "slop_master_table_out = '../output/' + today + '_RefSeq-ClinVar_GRCh37_slop-recovery-master.tsv'\n",
    "slop_master_table.to_csv(slop_master_table_out, sep='\\t', index=False)\n",
    "\n",
    "if os.path.isfile(slop_master_table_out):\n",
    "    print('\\nSuccess! File saved to:\\n{}'.format(tmp2_out))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
